File: 0350-intersection-of-two-arrays-ii/0350-intersection-of-two-arrays-ii.py
"""
class Solution:
    def intersect(self, nums1: List[int], nums2: List[int]) -> List[int]:
        occurrences = {}

        for n in nums1:
            if not occurrences.get(n):
                occurrences[n] = 1
            else:
                occurrences[n] += 1

        intersect = []

        for n in nums2:
            if occurrences.get(n, 0) > 0:
                intersect.append(n)
                occurrences[n] -= 1

        return intersect
"""

class Solution:
    def intersect(self, nums1: List[int], nums2: List[int]) -> List[int]:
        intersect = []

        nums1.sort()
        nums2.sort()

        i, j = 0, 0
        while i < len(nums1) and j < len(nums2):
            if nums1[i] < nums2[j]:
                i += 1
            elif nums2[j] < nums1[i]:
                j += 1
            else:
                intersect.append(nums1[i])    
                i += 1
                j += 1

        return intersect        

File: 0350-intersection-of-two-arrays-ii/0350-intersection-of-two-arrays-ii.go
/*
func intersect(nums1 []int, nums2 []int) []int {
    occurrences := map[int]int{}

    for _, v := range nums1 {
        occurrences[v]++
    }

    intersect := make([]int, 0, len(nums2))

    for _, v := range nums2 {
        if occ, ok := occurrences[v]; ok && occ > 0 {
            intersect = append(intersect, v)
            occurrences[v]--
        }
    }

    return intersect
}
*/

import (
    "sort"
)

func intersect(nums1 []int, nums2 []int) []int {
    sort.Ints(nums1)
    sort.Ints(nums2)

    intersect := make([]int, 0, len(nums1))

    i, j := 0, 0
    for i < len(nums1) && j < len(nums2) {
        if nums1[i] < nums2[j] {
            i++
        } else if nums2[j] < nums1[i] {
            j++
        } else {
            intersect = append(intersect, nums1[i])
            i++
            j++
        }
    }

    return intersect
}